src/PowerShellEditorServices/Services/PowerShell/Execution/SynchronousPowerShellTask.cs

```diff
@@ -266,8 +331,42 @@ private void StopDebuggerIfRemoteDebugSessionFailed()
             }
         }
 
-        private void CancelNormalExecution() => _pwsh.Stop();
+        private void CancelNormalExecution()
+        {
+            if (_pwsh?.Runspace?.RunspaceStateInfo?.IsUsable() is false)
+            {
+                return;
+            }
 
-        private void CancelDebugExecution() => _pwsh.Runspace.Debugger.StopProcessCommand();
+            // If we're signaled to exit a runspace then that'll trigger a stop,
+            // if we block on that stop we'll never exit the runspace (
+            // and essentially deadlock).
+            if (_frame?.SessionExiting is true)
+            {
+                _pwsh?.BeginStop(null, null);
+                return;
+            }
+
+            try
+            {
+                _pwsh?.Stop();
+            }
+            catch (NullReferenceException nre)
+            {
+                _logger.LogError(
+                    nre,
+                    "Null reference exception from PowerShell.Stop received.");
+            }
+        }
+
+        private void CancelDebugExecution()
+        {
+            if (_pwsh?.Runspace?.RunspaceStateInfo?.IsUsable() is false)
+            {
+                return;
+            }
+
+            _pwsh?.Runspace?.Debugger?.StopProcessCommand();
+        }
```

Comment:
Yeah, that doesn't look like hell at all 🙄

Reply:
Ah yeah I can probably ditch all the null conditionals here since it turned out that the NRE was coming from SMA for some reason.

Reply:
This definitely looked like the result of "wtf is throwing that!" 😆 So sorry you had to debug that!